nvbench/detail/measure_cupti.cu

```diff
@@ -0,0 +1,288 @@
+{
+  std::vector<std::string> metrics;
+  metrics.reserve(metric_ids::count);
+
+  add_metric<metric_ids::dram_peak_sustained_throughput>(state, metrics);
+  return metrics;
+}
+
+
+measure_cupti_base::measure_cupti_base(state &exec_state)
+  : m_state{exec_state}
+{
+  try
+  {
+    if (const std::optional<nvbench::device_info> &device = m_state.get_device();
+      device)
+    {
+      m_cupti = cupti_profiler(device, add_metrics(m_state));
+
+      if (!m_cupti->initialized())
+      {
+        m_cupti.reset();
+      }
+    }
+  }
+  catch (...)
+  {
+    m_cupti.reset();
+  }
+}
+
+
+void measure_cupti_base::check()
+{
+  if (!m_cupti)
+  {
+    if (auto printer_opt_ref = m_state.get_benchmark().get_printer();
+        printer_opt_ref.has_value())
+    {
+      auto &printer = printer_opt_ref.value().get();
+
+      printer.log(nvbench::log_level::fail,
+                  "CUPTI failed to collect metrics. "
+                  "Make sure you have permission to access NVIDIA GPU "
+                  "Performance Counters on the target device: "
+                  "https://developer.nvidia.com/"
+                  "nvidia-development-tools-solutions-err_nvgpuctrperm-"
+                  "permission-issue-performance-counters");
+    }
+  }
+
+  const auto device = m_state.get_device();
+  if (!device)
+  {
+    throw std::runtime_error(fmt::format("{}:{}: Device required for `cupti` "
+                                         "measurement.",
+                                         __FILE__,
+                                         __LINE__));
+  }
+  if (!device->is_active())
+  { // This means something went wrong higher up. Throw an error.
+    throw std::runtime_error(fmt::format("{}:{}: Internal error: Current "
+                                         "device is not active.",
+                                         __FILE__,
+                                         __LINE__));
+  }
+}
+
+
+template <metric_ids id>
+void gen_summary(std::size_t result_id,
+                 nvbench::state &m_state,
+                 const std::vector<double> &result)
+{
+  using metric = metric_info<id>;
+
+  if (metric::is_collected(m_state))
+  {
+    auto &summ = m_state.add_summary(metric::summary);
+    summ.set_string("hint", metric::hint);
+    summ.set_string("short_name", metric::short_name);
+    summ.set_string("description", metric::description);
+    summ.set_float64("value", result[result_id++] / metric::divider);
+  }
+
+  gen_summary<static_cast<metric_ids>(id + 1)>(result_id, m_state, result);
+}
+
+
+template <>
+void gen_summary<metric_ids::count>(std::size_t,
+                                    nvbench::state &,
+                                    const std::vector<double> &)
+{}
+
+
+void gen_summaries(nvbench::state &state,
+                   const std::vector<double> &result)
+{
+  gen_summary<metric_ids::dram_peak_sustained_throughput>(0, state, result);
+}
+
+
+void measure_cupti_base::generate_summaries()
+{
+  try
```

Comment:
Consider the function scope try/catch syntax when the entire function is in a try block:

```
void function()
try {
  ...
}
catch (...)
{
  ...
}
```